external/llvm-project/mlir/lib/Transforms/Utils/OutlinerUtils.cpp

```diff
@@ -173,337 +209,380 @@ static bool opsMatch(Operation *lhs, Operation *rhs, OutliningCandidate &one,
+
+class OutlineBuilder {
+
+  // Debug messaging for traversal
+  inline void debug(const char *tag, Operation *op) const {
+    LLVM_DEBUG(llvm::dbgs() << tag << ": " << op << "\n");
   }
-  for (auto params : llvm::zip(one.params, two.params)) {
-    if (std::get<0>(params) != std::get<1>(params)) {
-      return false;
+
+  inline void debug(const char *tag, Value v) const {
+    LLVM_DEBUG(llvm::dbgs() << tag << ": "; v.print(
+        llvm::dbgs(), OpPrintingFlags().elideLargeElementsAttrs());
+               llvm::dbgs() << "\n");
+  }
+
+  // Add Value/Operation methods
+  void addInput(Value v) {
+    debug("INPUT", v);
+    _inputs.insert(v);
+  }
+
+  void addResult(Value v) {
+    debug("RESULT", v);
+    _results.insert(v);
+  }
+
+  void addPredecessor(Operation *op) {
+    if (!_leadingOps.contains(op)) {
+      assert(!_trailingOps.contains(op));
+      debug("PRED", op);
+      _leadingOps.insert(op);
+      _locs.push_back(op->getLoc());
     }
   }
 
-  for (auto ops : llvm::zip(one.leadingOps, two.leadingOps)) {
-    if (!opsMatch(std::get<0>(ops), std::get<1>(ops), one, two)) {
-      return false;
+  void addSuccessor(Operation *op) {
+    if (!_trailingOps.contains(op)) {
+      debug("SUCC", op);
+      _trailingOps.insert(op);
+      _locs.push_back(op->getLoc());
     }
   }
-  if (!opsMatch(one.anchorOp, two.anchorOp, one, two)) {
-    return false;
+
+  bool contains(Operation *op) const {
+    return _trailingOps.contains(op) || _leadingOps.contains(op);
   }
-  for (auto ops : llvm::zip(one.trailingOps, two.trailingOps)) {
-    if (!opsMatch(std::get<0>(ops), std::get<1>(ops), one, two)) {
+
+  // Recurse back full chain looking for anchorOp
+  // - uses encountered set to reduce time
+  bool inChainToAnchor(Operation *inOp,
+                       DenseSet<Operation *> &encountered) const {
+    if (inOp == nullptr)
+      return false;
+    else if (inOp == _anchorOp)
+      return true;
+    else if (encountered.contains(inOp))
       return false;
+    encountered.insert(inOp);
+    for (auto opr : inOp->getOperands()) {
+      if (Operation *oprOp = opr.getDefiningOp()) {
+        debug(">>>", oprOp);
+        if (inChainToAnchor(oprOp, encountered))
+          return true;
+        debug("<<<", oprOp);
+      }
     }
+    return false;
   }
-  return true;
-}
 
-static OutliningCandidate *
-findOutliningCandidate(OutliningCandidate &newCandidate,
-                       std::vector<OutliningCandidate> &candidates) {
-  for (auto &candidate : candidates) {
-    if (outliningCandidatesEquivalent(candidate, newCandidate)) {
-      return &candidate;
+  // Make sure all operands are external or immediate in the chain from anchorOp
+  bool isImmediateOp(Operation *inOp) const {
+    for (auto opr : inOp->getOperands()) {
+      if (Operation *oprOp = opr.getDefiningOp()) {
+        debug("IN>", oprOp);
+        DenseSet<Operation *> encountered;
+        if (!contains(oprOp) && inChainToAnchor(oprOp, encountered))
```

Comment:
I believe, but haven't rigorously proven, that this inChainToAnchor() call is equivalent to domInfo.properlyDominates(anchor, oprOp);  thus more conservative than the old version and somewhat uglier, but otherwise okay.

Reply:
I walked thru that code, but it seemed to only check that they were in the same block and one op came after another.. It didn't check for any relation via operands.